File: sakinah/js/core/tasbih.js
/**
 * المسبحة الإلكترونية: منطق العدّ (نقي، بلا واجهة) — صيغ الذكر، الهدف، الدورات، وإحصاء اليوم والإجمالي لكل صيغة.
 * الحالة تُحفظ في الإعدادات (tasbih) فيستمر العدّ بعد إغلاق التطبيق.
 */
export const TASBIH_PHRASES = [
  { id: 'subhan', text: 'سُبْحَانَ اللهِ' },
  { id: 'hamd', text: 'الْحَمْدُ لِلهِ' },
  { id: 'akbar', text: 'اللهُ أَكْبَرُ' },
  { id: 'tahlil', text: 'لَا إِلَهَ إِلَّا اللهُ' },
  { id: 'istighfar', text: 'أَسْتَغْفِرُ اللهَ' },
  { id: 'hawqala', text: 'لَا حَوْلَ وَلَا قُوَّةَ إِلَّا بِاللهِ' },
  { id: 'subhanbihamd', text: 'سُبْحَانَ اللهِ وَبِحَمْدِهِ' },
  { id: 'subhanazim', text: 'سُبْحَانَ اللهِ الْعَظِيمِ' },
  { id: 'salawat', text: 'اللَّهُمَّ صَلِّ وَسَلِّمْ عَلَى نَبِيِّنَا مُحَمَّدٍ' },
  { id: 'baqiyat', text: 'سُبْحَانَ اللهِ، وَالْحَمْدُ لِلهِ، وَلَا إِلَهَ إِلَّا اللهُ، وَاللهُ أَكْبَرُ' },
  { id: 'custom', text: '' },
];
export const TASBIH_TARGETS = [33, 99, 100, 1000, 0]; // 0 = بلا حدّ

export function defaultTasbih() { return { phrase: 'subhan', target: 33, custom: '', count: 0, rounds: 0, totals: {} }; }
export function normalizeTasbih(s) { const d = defaultTasbih(); return s && typeof s === 'object' ? { ...d, ...s, totals: s.totals && typeof s.totals === 'object' ? s.totals : {} } : d; }
export function phraseText(state) {
  const p = TASBIH_PHRASES.find((x) => x.id === state.phrase) || TASBIH_PHRASES[0];
  return p.id === 'custom' ? (state.custom || '').trim() || 'ذكر' : p.text;
}
/** نقرة واحدة: تزيد العدّ وإحصاءات الصيغة؛ عند بلوغ الهدف تكتمل دورة ويعود العدّ إلى الصفر. تعيد الحالة الجديدة وهل اكتملت دورة */
export function tap(state, todayKey) {
  const s = { ...state, totals: { ...state.totals } };
  const key = s.phrase;
  const t = { ...(s.totals[key] || { total: 0, today: 0, date: todayKey }) };
  if (t.date !== todayKey) { t.today = 0; t.date = todayKey; }
  t.total += 1; t.today += 1; s.totals[key] = t;
  s.count += 1;
  let reached = false;
  if (s.target > 0 && s.count >= s.target) { reached = true; s.rounds += 1; s.count = 0; }
  return { state: s, reached };
}
/** تراجع عن نقرة (خطأ) */
export function undo(state, todayKey) {
  if (state.count <= 0 && state.rounds <= 0) return state;
  const s = { ...state, totals: { ...state.totals } };
  const t = s.totals[s.phrase]; if (t && t.total > 0) { const nt = { ...t }; nt.total -= 1; if (nt.date === todayKey && nt.today > 0) nt.today -= 1; s.totals[s.phrase] = nt; }
  if (s.count > 0) s.count -= 1; else { s.rounds -= 1; s.count = s.target > 0 ? s.target - 1 : 0; }
  return s;
}
export function resetCount(state) { return { ...state, count: 0, rounds: 0 }; }
export function todayCount(state, todayKey) { const t = state.totals[state.phrase]; return t && t.date === todayKey ? t.today : 0; }
export function totalCount(state) { const t = state.totals[state.phrase]; return t ? t.total : 0; }
/** إجمالي كل الصيغ (لبطاقة الإحصاءات) */
export function grandTotal(state) { return Object.values(state.totals || {}).reduce((n, t) => n + (t.total || 0), 0); }

File: sakinah-native/Sources/SakinahCore/Tasbih.swift
import Foundation

/// حالة المسبحة (تُحفظ كما في الويب): الصيغة، الهدف (0 = بلا حدّ)، نص مخصّص، العدّ، الدورات، وإحصاء كل صيغة
public struct TasbihState: Sendable, Hashable, Codable {
  public struct Totals: Sendable, Hashable, Codable { public var total: Int; public var today: Int; public var date: String }
  public var phrase: String = "subhan"
  public var target: Int = 33
  public var custom: String = ""
  public var count: Int = 0
  public var rounds: Int = 0
  public var totals: [String: Totals] = [:]
  public init() {}
  public init(from d: Decoder) throws {
    let c = try d.container(keyedBy: CodingKeys.self)
    phrase = (try? c.decode(String.self, forKey: .phrase)) ?? "subhan"; target = (try? c.decode(Int.self, forKey: .target)) ?? 33
    custom = (try? c.decode(String.self, forKey: .custom)) ?? ""; count = (try? c.decode(Int.self, forKey: .count)) ?? 0; rounds = (try? c.decode(Int.self, forKey: .rounds)) ?? 0
    totals = (try? c.decode([String: Totals].self, forKey: .totals)) ?? [:]
  }
}

public enum Tasbih {
  public static var phrases: [TasbihPhrase] { Catalog.shared.tasbih.phrases }
  public static var targets: [Int] { Catalog.shared.tasbih.targets }
  public static func phraseText(_ s: TasbihState) -> String {
    let p = phrases.first { $0.id == s.phrase } ?? phrases[0]
    if p.id == "custom" { let t = s.custom.trimmingCharacters(in: .whitespaces); return t.isEmpty ? "ذكر" : t }
    return p.text
  }
  /// نقرة: تزيد العدّ وإحصاء الصيغة؛ عند بلوغ الهدف تكتمل دورة ويعود العدّ صفرًا
  public static func tap(_ state: TasbihState, today: String) -> (state: TasbihState, reached: Bool) {
    var s = state
    var t = s.totals[s.phrase] ?? TasbihState.Totals(total: 0, today: 0, date: today)
    if t.date != today { t.today = 0; t.date = today }
    t.total += 1; t.today += 1; s.totals[s.phrase] = t
    s.count += 1
    var reached = false
    if s.target > 0 && s.count >= s.target { reached = true; s.rounds += 1; s.count = 0 }
    return (s, reached)
  }
  public static func undo(_ state: TasbihState, today: String) -> TasbihState {
    if state.count <= 0 && state.rounds <= 0 { return state }
    var s = state
    if var t = s.totals[s.phrase], t.total > 0 { t.total -= 1; if t.date == today && t.today > 0 { t.today -= 1 }; s.totals[s.phrase] = t }
    if s.count > 0 { s.count -= 1 } else { s.rounds -= 1; s.count = s.target > 0 ? s.target - 1 : 0 }
    return s
  }
  public static func reset(_ state: TasbihState) -> TasbihState { var s = state; s.count = 0; s.rounds = 0; return s }
  public static func todayCount(_ s: TasbihState, today: String) -> Int { guard let t = s.totals[s.phrase], t.date == today else { return 0 }; return t.today }
  public static func totalCount(_ s: TasbihState) -> Int { s.totals[s.phrase]?.total ?? 0 }
  public static func grandTotal(_ s: TasbihState) -> Int { s.totals.values.reduce(0) { $0 + $1.total } }
}
